File: csv/src/main/scala/akka/stream/alpakka/csv/scaladsl/CsvFormatting.scala
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.csv.scaladsl

import java.nio.charset.{Charset, StandardCharsets}

import akka.NotUsed
import akka.stream.alpakka.csv.impl.CsvFormatter
import akka.stream.scaladsl.{Flow, Source}
import akka.util.ByteString

import scala.collection.immutable

/** Provides CSV formatting flows that convert a sequence of String into their CSV representation
 * in [[akka.util.ByteString]].
 */
object CsvFormatting {

  val Backslash: Char = '\\'
  val Comma: Char = ','
  val SemiColon: Char = ';'
  val Colon: Char = ':'
  val Tab: Char = '\t'
  val DoubleQuote: Char = '"'

  /**
   * Create a Flow for converting iterables to ByteString.
   * @param endOfLine Line ending (default CR, LF)
   * @param quotingStyle Quote all fields, or only fields requiring quotes (default)
   * @param charset Character set, defaults to UTF-8
   * @param byteOrderMark Certain CSV readers (namely Microsoft Excel) require a Byte Order mark, defaults to None
   */
  def format[T <: immutable.Iterable[String]](
      delimiter: Char = Comma,
      quoteChar: Char = DoubleQuote,
      escapeChar: Char = Backslash,
      endOfLine: String = "\r\n",
      quotingStyle: CsvQuotingStyle = CsvQuotingStyle.Required,
      charset: Charset = StandardCharsets.UTF_8,
      byteOrderMark: Option[ByteString] = None
  ): Flow[T, ByteString, NotUsed] = {
    val formatter =
      new CsvFormatter(delimiter, quoteChar, escapeChar, endOfLine, quotingStyle, charset)
    byteOrderMark.fold {
      Flow[T].map(formatter.toCsv).named("CsvFormatting")
    } { bom =>
      Flow[T].map(formatter.toCsv).named("CsvFormatting").prepend(Source.single(bom))
    }

  }
}

File: csv/src/main/java/akka/stream/alpakka/csv/javadsl/CsvFormatting.java
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.csv.javadsl;

import akka.NotUsed;
import akka.stream.alpakka.csv.scaladsl.CsvQuotingStyle$;
import akka.stream.javadsl.Flow;
import akka.util.ByteString;
import scala.Option;
import scala.Some;
import scala.jdk.javaapi.CollectionConverters;
import scala.collection.immutable.List;

import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Collection;
import java.util.Optional;

/**
 * Provides CSV formatting flows that convert a sequence of String into their CSV representation in
 * {@see akka.util.ByteString}.
 */
public class CsvFormatting {

  public static final char BACKSLASH = '\\';
  public static final char COMMA = ',';
  public static final char SEMI_COLON = ';';
  public static final char COLON = ':';
  public static final char TAB = '\t';
  public static final char DOUBLE_QUOTE = '"';
  public static final String CR_LF = "\r\n";

  /**
   * Generates standard CSV format (with commas).
   *
   * @param <T> Any collection implementation
   * @return The formatting flow
   */
  public static <T extends Collection<String>> Flow<T, ByteString, NotUsed> format() {
    return format(
        COMMA,
        DOUBLE_QUOTE,
        BACKSLASH,
        CR_LF,
        CsvQuotingStyle.REQUIRED,
        StandardCharsets.UTF_8,
        Optional.empty());
  }

  /**
   * Generates CSV with the specified special characters and character set.
   *
   * @param delimiter Delimiter between columns
   * @param quoteChar Quoting character
   * @param escapeChar Escape character
   * @param endOfLine End of line character sequence
   * @param quotingStyle Quote all values or as required
   * @param charset Character set to be used
   * @param <T> Any collection implementation
   * @return The formatting flow
   */
  public static <T extends Collection<String>> Flow<T, ByteString, NotUsed> format(
      char delimiter,
      char quoteChar,
      char escapeChar,
      String endOfLine,
      CsvQuotingStyle quotingStyle,
      Charset charset,
      Optional<ByteString> byteOrderMark) {
    akka.stream.alpakka.csv.scaladsl.CsvQuotingStyle qs =
        CsvQuotingStyle$.MODULE$.asScala(quotingStyle);
    Option<ByteString> byteOrderMarkScala =
        byteOrderMark.<Option<ByteString>>map(Some::apply).orElse(Option.empty());
    akka.stream.scaladsl.Flow<List<String>, ByteString, NotUsed> formattingFlow =
        akka.stream.alpakka.csv.scaladsl.CsvFormatting.format(
            delimiter, quoteChar, escapeChar, endOfLine, qs, charset, byteOrderMarkScala);
    return Flow.<T>create()
        .map(c -> CollectionConverters.asScala(c).toList())
        .via(formattingFlow);
  }
}
